EHN: Add encoding_errors option in pandas.DataFrame.to_csv (#27750)

shigemk2 opened the pull request:
- [x] closes #27750
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -3151,6 +3152,12 @@ def to_csv(
         decimal : str, default '.'
             Character recognized as decimal separator. E.g. use ',' for
             European data.
+        errors : str, default 'strict'
```
The possible values should be listed here, like `{'strict', 'ignore', 'replace'}`.

TomAugspurger replied:
The table at https://docs.python.org/3/library/codecs.html#codec-base-classes has other options like xxmlcharrefreplace. Are those valid here?

TomAugspurger replied:
Can you also note that this only applies when pandas is the one opening the file?

We should validate that the user passing `errors` when they provide a filelike object, which we can't pass `errors` to, raises.

shigemk2 replied:
According to https://docs.python.org/3/library/codecs.html#codec-base-classes, surrogateescape is in 3.1, surrogatepass is in 3.4, namereplace and backslashreplace are in 3.5. So, may I add only `backslashreplace` option? Or should I add all error handling options?

shigemk2 replied:
Sorry, we can handle all error options because current pandas depends on Python 3.7.

TomAugspurger replied:
We require on Python 3.5.3 or newer.

jreback replied:
I know we like short kwargs, but I think ``encoding_errors`` would be better here

TomAugspurger replied:
Either is fine by me.

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -3151,6 +3152,12 @@ def to_csv(
             Character recognized as decimal separator. E.g. use ',' for
             European data.
+        errors : str, default 'strict'
+            Response when the input string can’t be converted according to
```
Change "Response" to "Behavior"?

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -3151,6 +3152,12 @@ def to_csv(
+        errors : str, default 'strict'
+            Response when the input string can’t be converted according to
+            the encoding’s rules
+            strict: Raise UnicodeError
```
This should be a list. (newline above and start each line with *).

TomAugspurger commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -514,3 +514,85 @@ def test_to_csv_compression(self, compression_only, read_infer, to_infer):
+        df = pd.DataFrame(data)
+        # case 1: errors=strict
+        with tm.ensure_clean("sjis_test.csv") as path:
+            expected_exception = UnicodeEncodeError
```
No need for this variable.

TomAugspurger commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -514,3 +514,85 @@ def test_to_csv_compression(self, compression_only, read_infer, to_infer):
+
+        # case 2: errors=replace
+        with tm.ensure_clean("sjis_test.csv") as path:
+            expected_sjis = b"int,str_sjis\n" b"1,abc\n" b"2,?\n" b"3,def\n"
```
Do you need multiple strings here on this line?

shigemk2 replied:
Do you mean `expected_sjis = "int,str_sjis\n1,abc\n2,?\n3,def\n"` ?

shigemk2 replied:
Sorry, you mean

```
            expected_cp932 = """\
int,str_cp932
1,abc
2,?
3,def
"""
```

TomAugspurger replied:
Either of those. Just shouldn't need the `b"ab\n"b"cd\n"b"ef\n".

TomAugspurger reviewed (commented):
Thanks. Can you add a release note to doc/source/whatsnew/v1.0.0? Under enhancements.

pep8speaks commented:
Hello @shigemk2! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`pandas/tests/io/formats/test_to_csv.py`](https://github.com/pandas-dev/pandas/blob/9f6d2073d489889b0623128c138137ce9f1abc6a/pandas/tests/io/formats/test_to_csv.py):

> [Line 586:89](https://github.com/pandas-dev/pandas/blob/9f6d2073d489889b0623128c138137ce9f1abc6a/pandas/tests/io/formats/test_to_csv.py#L586): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 676:89](https://github.com/pandas-dev/pandas/blob/9f6d2073d489889b0623128c138137ce9f1abc6a/pandas/tests/io/formats/test_to_csv.py#L676): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 88 characters)
> [Line 766:89](https://github.com/pandas-dev/pandas/blob/9f6d2073d489889b0623128c138137ce9f1abc6a/pandas/tests/io/formats/test_to_csv.py#L766): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 856:89](https://github.com/pandas-dev/pandas/blob/9f6d2073d489889b0623128c138137ce9f1abc6a/pandas/tests/io/formats/test_to_csv.py#L856): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 88 characters)

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -21,7 +21,7 @@ including other versions of pandas.
 ~~~~~~~~~~~~
 
--
+- Add errors option in pandas.DataFrame.to_csv (:issue:`27750`)
```
use ``:meth:`Dataframe.to_csv` ``, use double-backticks around ``errors``, say errors kwarg

shigemk2 commented:
@TomAugspurger @jreback 
Please review my PR.

TomAugspurger reviewed (commented):
Very thorough on the tests :)

Do you have time to add the new parameter to `user_guide/io.rst`? We repeat most of the parameters there, since there are so many. You can place it near `encoding` (~line 331). You likely just copy-paste the docstring description.

Can you also make it clear that this is the `errors` argument passed to :func:`\open\`?

jreback commented on pandas/core/generic.py:
```diff
@@ -3151,6 +3152,10 @@ def to_csv(
             Character recognized as decimal separator. E.g. use ',' for
             European data.
+        encoding_errors : str, default 'strict'
+            Behavior when the input string can’t be converted according to
```
needs a versionadded

jreback commented on pandas/io/common.py:
```diff
@@ -331,6 +337,10 @@ def _get_handle(
         mode (pickle, etc.)
+    encoding_errors : str, default 'strict'
+        Behavior when the input string can’t be converted according to
+        the encoding’s rules (strict, ignore, replace, etc.)
```
same

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -514,3 +514,463 @@ def test_to_csv_compression(self, compression_only, read_infer, to_infer):
             tm.assert_frame_equal(result, df)
+
+    def test_to_csv_path_with_sjis(self):
+        data = {"int": [1, 2, 3], "str_sjis": ["abc", "\u070a", "def"]}
```
as issue number as a comment

shigemk2 commented:
@TomAugspurger 

> Can you also make it clear that this is the errors argument passed to :func:\open\?

What do you mean?

TomAugspurger commented:
Include that explicit reference in the docstring. 

> On Aug 19, 2019, at 07:00, Michihito Shigemura <notifications@github.com> wrote:
> 
> @TomAugspurger
> 
> Can you also make it clear that this is the errors argument passed to :func:\open?
> 
> What do you mean?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

shigemk2 commented:
@TomAugspurger 

Do you mean to add to_csv's example in https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#writing-out-data ?

TomAugspurger commented:
I mean under https://dev.pandas.io/user_guide/io.html#quoting-compression-and-file-format

shigemk2 commented:
I add documents for error options both in  https://dev.pandas.io/user_guide/io.html#quoting-compression-and-file-format and https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#writing-out-data .

TomAugspurger commented on doc/source/user_guide/io.rst:
```diff
@@ -332,6 +332,11 @@ encoding : str, default ``None``
   Encoding to use for UTF when reading/writing (e.g. ``'utf-8'``). `List of
   Python standard encodings
   <https://docs.python.org/3/library/codecs.html#standard-encodings>`_.
+encoding_errors : str, default 'strict'
```
Sorry, I think I led you astray here. This should be removed, since this section is for `read_csv`, and the change is just for `to_csv`.

shigemk2 commented:
@TomAugspurger @jreback
Please review my PR again.

WillAyd commented:
Late to review but would we rather just document how to do this with the underlying file handle? Personally against adding new keywords to the read_* functions if they don't add functionality that can't already be done with the underlying fp object

TomAugspurger commented:
@WillAyd that was my initial reaction too. The thing that changed my mind was that we already accept the `encoding` and `mode` arguments.

Though I didn't realize `open` had so many arguments. With this change, we would (I think) support 4/8 of them.

WillAyd commented:
Yea certainly a grey area. If we are only at 4/8 I *think* we would ideally either commit to all of them or start pushing people to passing through directly to the file constructor. 

The latter of the two seems more reasonable though hence why I think documentation would be better (but not a hang up if you and other core devs feel strongly otherwise)

shigemk2 commented:
According to https://docs.python.org/3/library/functions.html#open, current(0.25.x) pandas already supports 5/8 open's aguments in _get_handle function. 

- [x] file
- [x] mode
- [ ] buffering
- [x] encoding
- [x] errors
- [x] newline
- [ ] closefd
- [ ] opener

https://github.com/pandas-dev/pandas/blob/0.25.x/pandas/io/common.py#L396-L405

https://github.com/pandas-dev/pandas/blob/3bbb6f3a634c851695514982bf10772fb5e2ff17/pandas/io/common.py#L418-L429

So, I think we should support errors parameter because _get_handle already uses errors argument.

shigemk2 commented:
@WillAyd @TomAugspurger
How do you think about my pull-request?

shigemk2 committed:
EHN: Add encoding_errors option in pandas.DataFrame.to_csv (#27750)

encoding_errors : str, default 'strict'
Behavior when the input string can’t be converted according to
the encoding’s rules (strict, ignore, replace, etc.)
See: https://docs.python.org/3/library/codecs.html#codec-base-classes

TomAugspurger commented:
I'm leaning towards accepting the PR. @WillAyd does that sound OK?

WillAyd commented:
While I think this PR in a nutshell is well done I think I'm still against the principal of adding this. For consistency wouldn't we then want to add to read_csv and other text based methods? Also do we see any concern with the fact that `errors` is the keyword argument to `open` but we now have `encoding_errors` in our API?

I just think to comprehensively support this would take a lot more effort than its worth so would rather just document how to do it instead of semi-supporting it via this method

TomAugspurger commented:
Thanks @WillAyd. All fair points.

In the past, we've come up against issues where we want to pass additional arguments through to the underlying filesystem calls. I wonder
if we can come up with a generic API here.

```
df.to_csv("foo.csv", storage_options={"mode", "w", "errors": "replace"})
```

Thoughts on that kind of API? It's a bit fuzzy on where exactly `storage_options` would go, but thoughts on the general approach?

WillAyd commented:
Maybe we just add kwargs and dictate that kwargs get passed through to the filepath_or_buffer argument? Not sure if that has been discussed in the past but seems natural

TomAugspurger commented:
Not sure if it's been discussed before. It's certainly possible. Which do you think is easier to document, `storage_options` or `**kwargs`? IMO, they're about the same...

Do any of our readers / writers already accept `**kwargs`?

shigemk2 commented:
@WillAyd @TomAugspurger 
IMO, we should talk about only `to_csv` in order not to go off the rails.
Of course, we have to talk about `raed_csv` (and `read_excel`) error handling.
But my PR does not have any impact on `read_csv`.

TomAugspurger commented:
We want to provide a consistent API for all our readers, though. We need to
design something that would work for
more than just to_csv.

On Thu, Sep 5, 2019 at 12:34 AM Michihito Shigemura <
notifications@github.com> wrote:

> @WillAyd <https://github.com/WillAyd> @TomAugspurger
> <https://github.com/TomAugspurger>
> IMO, we should talk about only to_csv in order not to go off the rails.
> Of course, we have to talk about raed_csv (and read_excel) error handling.
> But my PR does not have any impact on read_csv.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/27899?email_source=notifications&email_token=AAKAOIS2PXSECA33JNG2PLDQICK7LA5CNFSM4ILJ4OTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD554FOA#issuecomment-528204472>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIWXJUCANZKAW3VK4W3QICK7LANCNFSM4ILJ4OTA>
> .
>

jreback commented:
I like @TomAugspurger suggestion of a dict of ``storage_options``, composing keywords for ``encoding, errors, newline`` I think. would require some deprecation / translation period though. We are already doing this for compression options: https://github.com/pandas-dev/pandas/pull/26024.

We could accept a dict (simpler), or make a named tuple ``StorageOptions`` (and change the compression above to be the same).

WillAyd commented:
So as far as this PR goes I would suggest going to the documentation route, showing how you can pass this parameter to a file handle prior to a `read_csv` call. Then we can open up a follow up issue for the larger effort of `storage_options` generically in our IO. 

Does that sound good @shigemk2 ?

shigemk2 commented:
@WillAyd 
I do not know about how to handle `read_csv`'s error, so I want to close this PR.

TomAugspurger commented:
OK. I opened https://github.com/pandas-dev/pandas/issues/28377 as a followup.

Sorry this didn't end up making it in @shigemk2.
